### PR TITLE
Add FixConservativesAndComputePrims to GH+GRMHD

### PR DESCRIPTION
**Pull request by nilsdeppe**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/FixConservativesAndComputePrims.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        conserved_vars_ptr,
+    const gsl::not_null<Variables<hydro::grmhd_tags<DataVector>>*>
+        primitive_vars_ptr,
+    const grmhd::ValenciaDivClean::FixConservatives& fix_conservatives,
+    const EquationsOfState::EquationOfState<true, ThermodynamicDim>& eos) {
+  // Compute the spatial metric, inverse spatial metric, and sqrt{det{spatial
+  // metric}}. Storing the allocation or the result in the DataBox might
+  // actually be useful here, but unclear. We will need to profile.
+  Variables<
+      tmpl::list<gr::Tags::SpatialMetric<3>, gr::Tags::InverseSpatialMetric<3>,
```
Comment (yoonso0-0) [inline src/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/FixConservativesAndComputePrims.cpp:40]: include `PointwiseFunctions/GeneralRelativity/Tags.hpp`

File: src/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/FixConservativesAndComputePrims.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                 gr::Tags::SqrtDetSpatialMetric<>>>
+      temp_buffer{conserved_vars_ptr->number_of_grid_points()};
+  auto& spatial_metric = get<gr::Tags::SpatialMetric<3>>(temp_buffer);
+  gr::spatial_metric(make_not_null(&spatial_metric),
+                     get<gr::Tags::SpacetimeMetric<3>>(*conserved_vars_ptr));
+  auto& inverse_spatial_metric =
+      get<gr::Tags::InverseSpatialMetric<3>>(temp_buffer);
+  auto& sqrt_det_spatial_metric =
+      get<gr::Tags::SqrtDetSpatialMetric<>>(temp_buffer);
+  determinant_and_inverse(make_not_null(&sqrt_det_spatial_metric),
```
Comment (yoonso0-0) [inline src/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/FixConservativesAndComputePrims.cpp:50]: include `DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp`

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+SPECTRE_TEST_CASE(
+    "Unit.Evolution.Systems.GhValenciaDivClean.Subcell.FixConsAndComputePrims",
+    "[Unit][Evolution]") {
+  using System = grmhd::GhValenciaDivClean::System;
+
+  // Only use 1 grid point to see that we correctly flagged the point as being
+  // fixed. We're really only testing that the mutator calls the correct
+  // functions.
+  const size_t num_pts = 1;
```
Comment (yoonso0-0) [inline tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp:35]: include `cstddef`

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  tnsr::aa<DataVector, 3, Frame::Inertial> spacetime_metric{num_pts, 0.0};
```
Comment (yoonso0-0) [inline tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp:36]: include `DataStructures/Tensor/Tensor.hpp`

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    spacetime_metric.get(i, i) = 1.0;
+  }
+  tnsr::ii<DataVector, 3, Frame::Inertial> spatial_metric{num_pts, 0.0};
+  tnsr::II<DataVector, 3, Frame::Inertial> inverse_spatial_metric{num_pts, 0.0};
+  for (size_t i = 0; i < 3; ++i) {
+    spatial_metric.get(i, i) = spacetime_metric.get(i + 1, i + 1);
+    inverse_spatial_metric.get(i, i) = 1.0 / spatial_metric.get(i, i);
+  }
+  const Scalar<DataVector> sqrt_det_spatial_metric{
+      sqrt(get(determinant(spatial_metric)))};
```
Comment (yoonso0-0) [inline tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp:48]: include for math expressions

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  auto box = db::create<db::AddSimpleTags<
+      grmhd::ValenciaDivClean::Tags::VariablesNeededFixing,
+      typename System::variables_tag, typename System::primitive_variables_tag,
+      ::Tags::VariableFixer<grmhd::ValenciaDivClean::FixConservatives>,
+      hydro::Tags::EquationOfState<
+          std::unique_ptr<EquationsOfState::EquationOfState<true, 1>>>>>(
+      false, cons_vars,
+      typename System::primitive_variables_tag::type{num_pts, 1.0e-4},
+      variable_fixer,
+      std::unique_ptr<EquationsOfState::EquationOfState<true, 1>>{
```
Comment (yoonso0-0) [inline tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp:69]: include `memory`

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+          std::make_unique<EquationsOfState::PolytropicFluid<true>>(eos)});
+
+  using recovery_schemes = tmpl::list<
```
Comment (yoonso0-0) [inline tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp:72]: include `Utilities/TMPL.hpp`

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+      make_not_null(&box));
+
+  // Verify that the conserved variables were fixed
+  CHECK(db::get<grmhd::ValenciaDivClean::Tags::VariablesNeededFixing>(box));
+
+  // Manually do a primitive recovery and see that the values match what's in
+  // the DataBox.
+  typename System::primitive_variables_tag::type expected_prims{num_pts,
+                                                                1.0e-4};
+  grmhd::ValenciaDivClean::PrimitiveFromConservative<recovery_schemes, true>::
```
Comment (yoonso0-0) [inline tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp:85]: include for `ValenciaDivClean::PrimitiveFromConservative()`

**Review by yoonso0-0 — COMMENTED** [review]
Only `#include` things from me. Please squash :)

Q) Is this routine solely for atmosphere treatment, or does it work for handling **any** unreasonable set of conservative vars in terms of primitive recovery?

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp (hunk 4 above)
Comment (kidder) [reply tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp:36]: also DataVector

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+      typename System::primitive_variables_tag::type{num_pts, 1.0e-4},
+      variable_fixer,
+      std::unique_ptr<EquationsOfState::EquationOfState<true, 1>>{
+          std::make_unique<EquationsOfState::PolytropicFluid<true>>(eos)});
+
+  using recovery_schemes = tmpl::list<
+      grmhd::ValenciaDivClean::PrimitiveRecoverySchemes::KastaunEtAl>;
+  db::mutate_apply<grmhd::GhValenciaDivClean::subcell::
+                       FixConservativesAndComputePrims<recovery_schemes>>(
+      make_not_null(&box));
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp:76]: include gsl

**Review by kidder — CHANGES_REQUESTED** [review]

File: tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp (hunk 5 above)
Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/Subcell/Test_FixConservativesAndComputePrims.cpp:48]: `DataVector.hpp` covers this

**Review by nilsdeppe — COMMENTED** [review]
I've rebased and squashed. Thanks for the reviews :)

**Review by yoonso0-0 — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**